### DISCUSSION TITLE
I'm creating a layout-of-layout

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#7309] [Ideas]
It a plugin, WIP. I will post my design here for discussion, because I want my design work for all upcoming layouts.

---

  "title": "$:/plugins/linonetwo/lol",
  "author": "LinOnetwo",
  "description": "Tabs of Note, Whiteboard, Calendar..., split tabs in grid, launch pad for layouts, all in LOL: Layout-Of-Layout / Lay-Out-Launcher ",

![截屏2023-03-02 00 09 32](https://user-images.githubusercontent.com/3746270/222196600-12a8515d-7dfa-48ba-9562-ebda9941b855.png)

---

!! Tabbed Grid

We use [[gridstack.ts|https://github.com/gridstack/gridstack.js]] as entry point of this plugin. We create a widget with this npm package.

This widget ([[$:/plugins/linonetwo/lol/widgets/tabbed-grid/index.js]]) will create a container div, and all sub-widget will become children of this div, so become a grid.

Each grid should have multiple tabs, and when a tab is selected, the content will load a layout using the technique similar to `$:/core/ui/RootTemplate`:

```
<-- use list widget to forEach group and forEach tab ->
<$transclude
  tiddler={{{ [{$:/layouts/state}jsonget<groupIndex>,[tabs],<tabIndex>,[layout]has[text]] ~[[$:/plugins/linonetwo/lol/default-empty-layout]] }}}
  mode="inline"
/>
```

`$:/layouts/state` is a tiddler for multi-tab UI state's persistence:

```json
[
  // this level is for tab-groups (grid item, or you can call it splited windows)
  {
    "tabs": [
      // this level is for a single tab, which content is a layout
      {
        "layout": "$:/plugins/linonetwo/tw-calendar/tiddlywiki-ui/PageLayout/CalendarLayout"
      },
      {
        "layout": "$:/core/ui/PageTemplate"
      }
    ]
  },
  {
    "tabs": {
      "layout": "$:/plugins/linonetwo/common-layouts/sidebar"
    }
  }
]
```

Use a list widget to generate tab-group, inside our `<$tabbed-grid>` widget.

---

TBD

**Comment by saqimtiaz** [comment]
@linonetwo that is interesting. I do something slightly similar with a custom widget that creates columns and rows that are resizable (though without drag and drop re-arrangement) loosely [inspired by this](https://storybook-collapse-pane.netlify.app/?path=/story/collapsable-panes--vertical-split) but with the facility to dynamically add and remove panels.

I would encourage you to think of the content of any tab being any tiddler and not just layouts. If need be, use a custom tag for such tiddlers distinct from $:/tags/Layout.

 Layouts are after all just tiddlers that are designed to be rendered by the root widget. However, in some situations layouts might not be appropriate content for tabs but other tiddlers might be. For example, you would not want to use two different layouts at the same time if both contain the same default sidebar.
 
 Similarly there might be tiddlers that appropriate to be used in the tabs but are not as useful used as templates for the root widget. Distinction between what goes into a tab and Layouts would there be useful.

**Reply by linonetwo** [reply]
So I was asking a fullscreen editor layout, which only renders one tiddler, but is still a layout

https://talk.tiddlywiki.org/t/is-there-a-fullscreen-edit-view-mode-or-layout/6235?u=linonetwo

And about sidebar, in the state json above, there is a sidebar layout that only shows sidebar, so other layouts can hide the sidebar by default.

One difficulty is how to pass event between them, like when click on a link in sidebar, how to change fullscreen editor layout's content.

**Reply by linonetwo** [reply]
I feel layouts are like fullscreen Apps on Android and iOS. While these os also support split screen to show two apps at once.

And single tiddler is a card content in the app.

**Reply by saqimtiaz** [reply]
> I feel layouts are like fullscreen Apps on Android and iOS. While these os also support split screen to show two apps at once.
> 
> And single tiddler is a card content in the app.

Except in TiddlyWiki everything is a tiddler. A layout is just a tiddler that contains a template that shows content from other tiddlers, and shows up in the layout switcher.

Personally for a template that renders the entire TW interface I use the term Layout, and for other templates that render a part of the entire interface, I use the term View with the tag $:/tags/Views. Some Views are usable as Layouts, and some Layouts are usable as Views. However, not all Views are useful as Layouts.

So this helps avoid the situation where the layout switcher gets cluttered with tiddlers that are not appropriate to be used as the entire TW interface.

**Reply by saqimtiaz** [reply]
> One difficulty is how to pass event between them, like when click on a link in sidebar, how to change fullscreen editor layout's content.

One approach would be to set a state tiddler that shows whether the sidebar is hidden or shown and allow the other layouts react to this change. The alternative that comes to mind is to send a widget message when opening/closing the sidebar that is handled by the container widget, which updates other layouts.

**Reply by linonetwo** [reply]
> $:/tags/Views

Well, tw core doesn't have this system tag. But you are right about

> So this helps avoid the situation where the layout switcher gets cluttered with tiddlers that are not appropriate to be used as the entire TW interface.

I think I can allow adding any tiddler to the launcher. Maybe by adding tag `$:/tags/Views`, or by add them to the `list` of a config tiddler.

**Reply by linonetwo** [reply]
> One difficulty is how to pass event between them, like when click on a link in sidebar, how to change fullscreen editor layout's content.

By this I mean, what's opened in story flow is controlled by a `$:/StoryList` tiddler. But when you have multiple layout side by side, you want them to have different StoryList. And click on a link on the sidebar should always open tiddler on left most one of layout (when you have two side by side).

So I might need to use `$:/layouts/state` to record StoryList of each layout, and find a way to inject the state to them.

**Reply by saqimtiaz** [reply]
> And click on a link on the sidebar should always open tiddler on left most one of layout (when you have two side by side).

You should be able to wrap the sidebar in a navigator widget and set the storylist and historylist to the value of a state tiddler. This state tiddler should always hold the storylist title and historylist title for the leftmost layout. Then the navigator widget will automatically update the correct storylist.

**Comment by linonetwo** [comment]
Not very successful, Demo:  https://tiddly-gittly.github.io/lol/

Switch to this layout:

![截屏2023-03-05 21 30 50](https://user-images.githubusercontent.com/3746270/222963479-df18b90e-f6a5-4d0b-a5df-7c6213a64ff0.png)


Problems encountered:

1. position: fix style make some element leak to cover other layouts, for example, default layout on the middle, will have its sidebar leak to cover the right most layout
1. need to design a ui to create and delete grid item, while not covering too much of space. For example, use a tab style handle
1. drag event should be minimized to a handle, for example, to a tab style handle

Code: https://github.com/tiddly-gittly/lol

**Reply by saqimtiaz** [reply]
I think this further supports the idea that using Layouts designed for the entire screen inside smaller portions of the viewport will be problematic.

The way that I would approach this is to create a "Dynamic layout" and inside this layout provide standard components like sidebar and maybe search. Then allow creation of different columns or rows inside the rest of the layout, each of which shows a View. The Views can be designed to be self-isolated and not impact other parts of the layout.

Another concern is going be that with many Views open at the same time, the refresh performance will be impacted. So at least for the tabs not visible, refresh should be suspended (as it is with the core tabs macro).

**Reply by linonetwo** [reply]
Yes, maybe not involving layout, just allow add tiddler to the grid. And some tiddler may have `<$calendar>` or `<$whiteboard` widget. Or just a storyflow macro call, or a sidebar macro call.
